### PR TITLE
VariantsSparkSink

### PR DESCRIPTION
**Pull request by tomwhite**
Addresses #1989. Note that this requires https://github.com/HadoopGenomics/Hadoop-BAM/pull/111 from Hadoop-BAM, so it's not ready to be merged yet.

@lbergelson would you like to take a look to see if the direction looks OK to you?

### CODE REVIEW

**Comment by tomwhite** [comment]
@lbergelson the new Hadoop-BAM is available for this, so can you review please?

File: src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/VariantsSparkSink.java [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+    // Output format class for writing VCF files through saveAsNewAPIHadoopFile. Must be public.
+    public static class SparkVCFOutputFormat extends KeyIgnoringVCFOutputFormat<NullWritable> {
+        public static VCFHeader vcfHeader;
+
+        public static void setVCFHeader(final VCFHeader header) {
+            vcfHeader = header;
+        }
+
+        public SparkVCFOutputFormat() {
+            super(VCFFormat.VCF);
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/VariantsSparkSink.java]: I assume this does not work with BCFs, correct?

File: src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/VariantsSparkSink.java [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+     * @param ctx the JavaSparkContext
+     * @param outputFile path to the output VCF
+     * @param variants variants to write
+     * @param header the header to put at the top of the output file
+     * @throws IOException if an error occurs while writing
+     */
+    public static void writeVariants(
+            final JavaSparkContext ctx, final String outputFile, final JavaRDD<VariantContext> variants,
+            final VCFHeader header) throws IOException {
+        String absoluteOutputFile = makeFilePathAbsolute(outputFile);
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/VariantsSparkSink.java]: Will this work with HDFS URIs?

File: src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/VariantsSparkSink.java [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+    public static class SparkHeaderlessVCFOutputFormat extends SparkVCFOutputFormat {
+        @Override
+        public RecordWriter<NullWritable, VariantContextWritable> getRecordWriter(TaskAttemptContext ctx) throws IOException {
+            ctx.getConfiguration().setBoolean(WRITE_HEADER_PROPERTY, false);
+            return super.getRecordWriter(ctx);
+        }
+    }
+
+    /**
+     * Write variants to the given output file in VCF format with the given header. Note that writing sharded output is not supported.
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/VariantsSparkSink.java]: Should we make writing sharded vcf output a ticket?

File: src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/VariantsSparkSink.java [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+            variantContextWritable.set(variantContext);
+            return new Tuple2<>(variantContext, variantContextWritable);
+        });
+    }
+
+    private static Class<? extends OutputFormat<NullWritable, VariantContextWritable>> getOutputFormat(final boolean writeHeader) {
+        return writeHeader ? SparkVCFOutputFormat.class : SparkHeaderlessVCFOutputFormat.class;
+    }
+
+    private static String makeFilePathAbsolute(String path){
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/VariantsSparkSink.java]: Can we move this to an appropriate utils class?

Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/VariantsSparkSink.java]: These are proliferating.  Can we put this in BucketUtils or IOUtils?  There's something similar in `FeatureInput.makeIntoAbsolutePath` that does something special for TILEDB uris.

File: src/test/java/org/broadinstitute/hellbender/engine/spark/datasources/VariantsSparkSinkUnitTest.java [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+        MiniClusterUtils.stopCluster(cluster);
+    }
+
+
+    @DataProvider(name = "loadVariants")
+    public Object[][] loadVariants() {
+        return new Object[][]{
+                {hg19_chr1_1M_dbSNP, ".vcf"},
+                {hg19_chr1_1M_dbSNP, ".vcf.bgz"},
+                {hg19_chr1_1M_dbSNP_modified, ".vcf"},
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/engine/spark/datasources/VariantsSparkSinkUnitTest.java]: Are these test vcfs large enough to be divided into multiple fragments?

File: src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/VariantsSparkSink.java (hunk 4 above)
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/VariantsSparkSink.java]: Should this try to make hdfs and file uri's absolute as well?

File: src/test/java/org/broadinstitute/hellbender/engine/spark/datasources/VariantsSparkSinkUnitTest.java [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+    private VCFHeader getHeader(String vcf) {
+        final java.nio.file.Path vcfPath = asPath(vcf);
+        try (SeekableStream stream = new SeekablePathStream(vcfPath)) {
+            return VCFHeaderReader.readHeaderFrom(stream);
+        } catch (IOException e) {
+            throw new UserException("Failed to read VCF header from " + vcf + "\n Caused by:" + e.getMessage(), e);
+        }
+    }
+
+    private static java.nio.file.Path asPath(String path) {
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/engine/spark/datasources/VariantsSparkSinkUnitTest.java]: Can you move this one to a utils class as well? Could be helpful elsewhere.

File: src/test/java/org/broadinstitute/hellbender/engine/spark/datasources/VariantsSparkSinkUnitTest.java [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+        URI uri = URI.create(path);
+        return uri.getScheme() == null ? Paths.get(path) : Paths.get(uri);
+    }
+
+    private List<VariantContext> readVariants(String vcf) throws IOException {
+        File actualVcf;
+        // work around TribbleIndexedFeatureReader not reading header from .bgz files
+        if (vcf.endsWith(".bgz")) {
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/engine/spark/datasources/VariantsSparkSinkUnitTest.java]: `.vcf.gz` is the more common extension for compressed vcfs -- should handle that as well.

**Comment by droazen** [comment]
Review complete, back to @tomwhite. Looks good, merge after addressing comments.

**Comment by lbergelson** [comment]
I think this should also fix #2119 since it includes https://github.com/HadoopGenomics/Hadoop-BAM/pull/114.

File: src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/VariantsSparkSink.java (hunk 1 above)
Comment (tomwhite) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/VariantsSparkSink.java]: Correct. Added a comment to the class doc.

File: src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/VariantsSparkSink.java (hunk 2 above)
Comment (tomwhite) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/VariantsSparkSink.java]: Yes, similar to code in ReadsSparkSink.

File: src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/VariantsSparkSink.java (hunk 3 above)
Comment (tomwhite) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/VariantsSparkSink.java]: Opened #2161.

File: src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/VariantsSparkSink.java (hunk 4 above)
Comment (tomwhite) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/VariantsSparkSink.java]: Moved to BucketUtils.

File: src/test/java/org/broadinstitute/hellbender/engine/spark/datasources/VariantsSparkSinkUnitTest.java (hunk 5 above)
Comment (tomwhite) [inline src/test/java/org/broadinstitute/hellbender/engine/spark/datasources/VariantsSparkSinkUnitTest.java]: Good idea. It's using one partition, so I've expanded the test to repartition.

File: src/test/java/org/broadinstitute/hellbender/engine/spark/datasources/VariantsSparkSinkUnitTest.java (hunk 6 above)
Comment (tomwhite) [inline src/test/java/org/broadinstitute/hellbender/engine/spark/datasources/VariantsSparkSinkUnitTest.java]: Done.

**Pull request merged**